lib/growl.js

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * @summary
+ * Checks if Growl support seems likely.
+ *
+ * @description
+ * Glosses over the distinction between an unsupported platform
+ * and one that lacks prerequisite software installations.
+ * Autofails when run in browser.
+ *
+ * @public
+ * @see {@link https://github.com/tj/node-growl/blob/master/README.md|Prerequisite Installs}
+ * @see {@link Mocha#growl}
+ * @see {@link Mocha#isGrowlCapable}
+ * @return {boolean} whether Growl support can be expected
+ */
+exports.isCapable = function() {
+  return !process.browser && which(getSupportBinaries()) !== '';
+};
+
+/**
+ * Implements desktop notifications as a pseudo-reporter.
+ *
+ * @public
+ * @see {@link Mocha#_growl}
+ * @param {Runner} runner - Runner instance.
+ */
+exports.notify = function(runner) {
+  var sendNotification = function() {
+    var growl = require('growl');
+    var stats = runner.stats;
+    var msg;
+    var options;
+
+    if (stats.failures) {
+      msg = stats.failures + ' of ' + runner.total + ' tests failed';
+      options = {
+        name: 'mocha',
+        title: 'Failed',
+        image: image('error')
+      };
+    } else {
+      msg = stats.passes + ' tests passed in ' + stats.duration + 'ms';
+      options = {
+        name: 'mocha',
+        title: 'Passed',
+        image: image('ok')
+      };
+    }
+    growl(msg, options, onCompletion);
+  };
+
+  /**
+   * @summary
+   * Callback for result of attempted Growl notification.
+   *
+   * @description
+   * Despite its appearance, this is <strong>not</strong> an Error-first
+   * callback -- all parameters are populated regardless of success.
+   *
+   * @callback Growl~growlCB
+   * @param {*} err - Error object, or <code>null</code> if successful.
+   * @param {string} stdout - <code>stdout</code> from notification delivery
+   *     process.
+   * @param {string} stderr - <code>stderr</code> from notification delivery
+   *     process. It will include timestamp and executed command with arguments.
+   */
+
+  function onCompletion(err, stdout, stderr) {
+    if (err) {
+      var detail =
+        err.code === 'ENOENT' ? 'prerequisite software not found' : err.message;
+      console.error('Growl notification error:', detail);
+    }
+  }
+
+  runner.once('end', sendNotification);
+};
+
+/**
+ * Returns Growl image `name` path.
+ *
+ * @private
+ * @param {string} name - Basename of associated Growl image.
+ * @return {string} Pathname to Growl image
+ */
+function image(name) {
+  return path.join(__dirname, '..', 'assets', 'growl', name + '.png');
+}
+
+/**
+ * @summary
+ * Locates a binary in the user's `PATH`.
+ *
+ * @description
+ * Takes a list of command names and searches the path for each executable
+ * file that would be run had these commands actually been invoked.
+ *
+ * @private
+ * @param {string[]} binaries - Names of binary files to search for.
+ * @return {string} absolute path of first binary found, or empty string if none
+ */
+function which(binaries) {
```

Comment:
would strongly prefer that if we need this functionality, we get it from a purpose-built module, e.g. [which](https://npm.im/which).

Reply:
I mean, there are cases here that we're not considering including permissions and execution bits... would rather just have that all taken care of than have to implement it ourselves.

Hand-rolling stuff like this is why our `--watch` functionality is such a mess.

Reply:
The **only** thing being done here is to check if _something_ is installed. TJ's package will handle **all** other issues when it spawns the notifier on our behalf.

Since all the notification binaries are put there via installers, permissions and such are _very_ unlikely to be a problem anyway.